# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [10]:
import Toolbox.apriorimining
# Wouter Mobach - s44465733
# Victor van der Gevel - s4630807


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `associationmining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [13]:
import sys
sys.path.append('./Toolbox')
from apriorimining import generate_association_rules
generate_association_rules()


Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath/filename and extension: Data/courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.0
set= { 6, 7 },  sup= 83.0
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
--------------------------------------------------------


*Double click to type your answer for 5.1.2 here:*
100% of students took course 6, physics
83.33% of students took course 2, math
83.33% of students took course 7, chemistry
There is a 83% association with courses 2 and 6, math and physics, these courses are often taken together.
There is a 83% association with courses 6 and 7, physics and chemistry, these courses are often taken together.




  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? You can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [14]:
import sys
sys.path.append('./Toolbox')
from apriorimining import generate_association_rules
generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 50
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath/filename and extension: Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: y
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 50.0
set= { Fargo (1

*Double click to type your answer for 5.2.2 here:*
The associations with the strongest confidence are for the movie Star Wars. For instance in rule 228 there is the highest confidence of 99.66 that people that have purchased 'Empire Stikes Back', 'Raiders of the Lost Ark' and 'The Return of the Jedi' will also purchase 'Star Wars'. These associations make sense because 'Empire Stikes Back' and 'The Return of the Jedi' are both Star Wars movies so someone who is interested in Star Wars will be likely to purchase all movies related to the franchise. Furthermore, all four of the movies in the association rule are very popular so they are likely to be purchased.


#### 5.2.3 
(1 point) Which movie has been bought by the most users? There are only few rules with more than three items. Why?

*Double click to type your answer for 5.2.3 here:*
The movie that has been bought the most is 'Star Wars' with support = 61.82. 
There is only one rule with more than three items because people dont tend to buy more than three items.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

*Double click to type your answer for 5.2.4 here:*
Most rules with a very high confidence have a support in range of 30 - 35. This range or a higher one can also be seen in rules with a lower confidence. Thus it is possible for itemsets to have a relatively low support but still have a high confidence.

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6 or https://en.wikipedia.org/wiki/Lift_(data_mining)) for the association pattern $ \text{ {a } } \rightarrow   \text{{b }} $. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

4. (1.5 points) Prove that if the confidence of the rule $ \text{ {a } } \rightarrow   \text{{b }} $ is less than the support of $ \text{ {b }}$  then
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> c(  \text{ {a } } \rightarrow   \text{{b }})
$$
and
$$
c(  \text{ {~a } } \rightarrow   \text{{b }})> s( {b })
$$

where $ \text{c(.)}$  denotes the rule confidence and  $ \text{s(.)}$ denote the support of an itemset. 

Hint: To prove the statement rewrite the confidence and the support of the rule in terms of probabilities: 
$$
c(  \text{ {a } } \rightarrow   \text{{b }})=  \frac{P( {a,b })}{P( {a })}
$$

and

$$
s( {b })=P( {b })
$$

Further more, make use of probability rules such as $p(b) = p(\text{~}a,b) + p(a,b)$ and $p(\text{~}a) = 1 - p(a)$ and algebraic rules such as $$ \frac{p(a,b)}{p(b)} < p(a) \rightarrow 1 - \frac{p(a,b)}{p(b)} > 1 - p(a) $$


*Double click to type your answer for 5.3.1 here:*
Please click on this field, otherwise it seems a bit rubbish.
1. 
confidence = P(a,b)/P(a) = 0.3/0.45 = 2/3
p = 0.8
p > confidence. So the rule is uninteresting.

2.
interest measure = P(b|a)/P(b) = (2/3) / 0.8 = 0.83333333333
interest measure < 1. Therefore it is negatively associated.

3.High confidence rules may not (always) be interesting, it was not in this instance. In this case the Lift was negatively ascociated. 
4.
1) Suppose:
c({a} -> {b}) = (P(a,b)/ P(a)) < P(b), which implies P(a)*P(b) > P(a,b).

c({~a} -> {b}) = P(~a,b)/ P(~a) = (P(b) − P(a,b)) / (1 − P(a)) 

Thus, we can write: 
c({~a} -> {b}) − c({a}) -> {b}) = (P(b) − P(a,b)) / (1 − P(a)) - (P(a,b)/ P(a)) = (P(a)*P(b) − P(a,b))/(P(a)*(1 − P(a))

which is positive because,  P({a})P({b}) > P({a,b}

2)
c({~a} −> {b}) − s(b) = (P(b) − P(a,b)) / (1 − P(a)) - P(b) = (P(a)*P(b) − P(a,b)) / (1 − P(a))

This is positive because P(a)*P(b) > P(a,b)


#### 5.3.2

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



*Double click to type your answer for 5.3.2 here:
Please click on this field, otherwise it seems a bit rubbish.

Total:
odd ratio = (105/87)/(40/62) = ~ 1.87069
ϕ = ((105/294) - (192/294) (145/294))/(root ((192/294) (102/294) (145/294) (149/294))) = ~ 0.14730
interest = (105/294) / 145 = ~ 0.00246

Students:
odd ratio = (2/9)/(5/20) = 8/9 (~0.88889)
ϕ = ((2/36) - (11/36) (7/36))/(root ((11/36) (25/36) (7/36) (29/36))) = ~ -0.02116
interest = (2/36) / 7 = ~ 0.00794

Adults:
odd ratio = (103/78)/(35/42) = ~ 1.58462
ϕ = ((103/258) - (181/258) (138/258))/(root ((181/258) (77/258) (138/258) (120/258))) = ~ 0.10506
interest = (103/258) / 138 = ~ 0.00289

When data is pooled together the odd ratio and the ϕ are both higher than the seperate data. The interest is lower when the data is pooled together.

